adapters/beachfront/beachfront.go

```diff
@@ -0,0 +1,541 @@
+	Ip             string           `json:"ip"`
+}
+
+type BeachfrontSlot struct {
+	Slot     string           `json:"slot"`
+	Id       string           `json:"id"`
+	Bidfloor float64          `json:"bidfloor"`
+	Sizes    []BeachfrontSize `json:"sizes"`
+}
+
+type BeachfrontSize struct {
+	W uint64 `json:"w"`
+	H uint64 `json:"h"`
+}
+
+type BeachfrontResponseSlot struct {
+	CrID  string  `json:"crid"`
+	Price float64 `json:"price"`
+	W     uint64  `json:"w"`
+	H     uint64  `json:"h"`
+	Slot  string  `json:"slot"`
+	Adm   string  `json:"adm"`
+}
+
+func (a *BeachfrontAdapter) Name() string {
+	return "beachfront"
+}
+
+func (a *BeachfrontAdapter) SkipNoCookies() bool {
+	return true
+}
+
+func (a *BeachfrontAdapter) MakeRequests(request *openrtb.BidRequest) ([]*adapters.RequestData, []error) {
+	errs := make([]error, 0, len(request.Imp))
+
+	var beachfrontRequests BeachfrontRequests
+	var reqJSON []byte
+
+	a.URI = func() string {
+		for i := range request.Imp {
+			if request.Imp[i].Video != nil {
+				// If there are any video imps, we will be running a video auction
+				// and dropping all of the banner actions.
+				return VideoEndpoint
+			}
+		}
+		return BannerEndpoint
+	}()
+
+	beachfrontRequests, err := preprocess(request, a.URI)
+	if err != nil {
+		errs = append(errs, err)
+		return nil, errs
+	}
+
+	if a.URI == VideoEndpoint {
+		reqJSON, err = json.Marshal(beachfrontRequests.Video)
+		a.URI = a.URI + beachfrontRequests.Video.AppId + VideoEndpointSuffix
+	} else {
+		reqJSON, err = json.Marshal(beachfrontRequests.Banner)
+	}
+
+	if err != nil {
+		errs = append(errs, err)
+		return nil, errs
+	}
+
+	headers := http.Header{}
+	headers.Add("Content-Type", "application/json;charset=utf-8")
+	headers.Add("Accept", "application/json")
+
+	if a.URI == BannerEndpoint {
+		if request.User != nil {
+			headers.Add("Cookie", "UserID="+request.User.ID+"; __io_cid="+request.User.BuyerUID)
+		}
+	}
+
+	return []*adapters.RequestData{{
+		Method:  "POST",
+		Uri:     a.URI,
+		Body:    reqJSON,
+		Headers: headers,
+	}}, errs
+}
+
+/*
+We have received a prebid request. It needs to be converted to a beachfront request. This is complicated
+by the fact that we have different servers for video/display and they have different contracts.
+*/
+func preprocess(req *openrtb.BidRequest, uri string) (BeachfrontRequests, error) {
+	var beachfrontReqs BeachfrontRequests
+	var err error
+
+	if uri == BannerEndpoint {
+		beachfrontReqs.Banner, err = getBannerRequest(req)
+		if err != nil {
+			return beachfrontReqs, err
+		}
+	} else {
+		beachfrontReqs.Video, err = getVideoRequest(req)
+		if err != nil {
+			return beachfrontReqs, err
+		}
+	}
+
+	return beachfrontReqs, err
+}
+
+func getBannerRequest(req *openrtb.BidRequest) (BeachfrontBannerRequest, error) {
+	var beachfrontReq BeachfrontBannerRequest
+	dec := json.NewDecoder(strings.NewReader(beachfrontBannerRequestTemplate))
+	for {
+		if err := dec.Decode(&beachfrontReq); err == io.EOF {
```

Comment:
It all that `beachfrontBannerRequestTemplate` is used for is populating default values into `beachfrontReq`, wouldn't it be better to simply either define a default instance of the struct, and then deep copy it into `beachfrontReq`, or define a `defaultBannerReq()` function to allocate the struct and fill in the non null default entries? Saves having to process a string through the JSON decoder for every request.